tools/gfx/cuda/render-cuda.cpp

```diff
@@ -0,0 +1,1506 @@
+            entryPointLayouts.add(new CUDAShaderObjectLayout(
+                programLayout->getEntryPointByIndex(i)->getTypeLayout()));
+        }
+
+    }
+
+    int getKernelIndex(UnownedStringSlice kernelName)
+    {
+        for (int i = 0; i < (int)programLayout->getEntryPointCount(); i++)
+        {
+            auto entryPoint = programLayout->getEntryPointByIndex(i);
+            if (kernelName == entryPoint->getName())
+            {
+                return i;
+            }
+        }
+        return -1;
+    }
+
+    void getKernelThreadGroupSize(int kernelIndex, UInt* threadGroupSizes)
+    {
+        auto entryPoint = programLayout->getEntryPointByIndex(kernelIndex);
+        entryPoint->getComputeThreadGroupSize(3, threadGroupSizes);
+    }
+};
+
+class CUDAShaderObject : public ShaderObject
+{
+public:
+    RefPtr<MemoryCUDAResource> bufferResource;
+    RefPtr<CUDAShaderObjectLayout> layout;
+    List<RefPtr<CUDAShaderObject>> objects;
+    List<RefPtr<CUDAResourceView>> resources;
+
+    virtual SlangResult init(Renderer* renderer, CUDAShaderObjectLayout* typeLayout);
+
+    virtual slang::TypeLayoutReflection* getElementTypeLayout() override
+    {
+        return layout->typeLayout;
+    }
+    virtual Slang::Index getEntryPointCount() override { return 0; }
+    virtual ShaderObject* getEntryPoint(Slang::Index index) override { return nullptr; }
+    virtual SlangResult setData(ShaderOffset const& offset, void const* data, size_t size)
+    {
+        size = Math::Min(size, bufferResource->getDesc().sizeInBytes - offset.uniformOffset);
+        SLANG_CUDA_RETURN_ON_FAIL(cudaMemcpy(
+            (uint8_t*)bufferResource->m_cudaMemory + offset.uniformOffset,
+            data,
+            size,
+            cudaMemcpyHostToDevice));
+        return SLANG_OK;
+    }
+    virtual SlangResult getObject(ShaderOffset const& offset, ShaderObject** object)
+    {
+        auto subObjectIndex =
+            layout->m_bindingRanges[offset.bindingRangeIndex].baseIndex + offset.bindingArrayIndex;
+        if (subObjectIndex >= objects.getCount())
+        {
+            *object = nullptr;
+            return SLANG_OK;
+        }
+        *object = objects[subObjectIndex].Ptr();
+        return SLANG_OK;
+    }
+    virtual SlangResult setObject(ShaderOffset const& offset, ShaderObject* object)
+    {
+        auto subObjectIndex =
+            layout->m_bindingRanges[offset.bindingRangeIndex].baseIndex + offset.bindingArrayIndex;
+        SLANG_ASSERT(
+            offset.uniformOffset ==
+            layout->m_bindingRanges[offset.bindingRangeIndex].uniformOffset +
+                offset.bindingArrayIndex * sizeof(void*));
+        auto cudaObject = dynamic_cast<CUDAShaderObject*>(object);
+        if (subObjectIndex >= objects.getCount())
+            objects.setCount(subObjectIndex + 1);
+        objects[subObjectIndex] = cudaObject;
+        return setData(offset, &cudaObject->bufferResource->m_cudaMemory, sizeof(void*));
```

Comment:
I honestly don't know the answer to this: are we guaranteed that the pointer size always matches between a CUDA device and the host program? Are there cases where we could have a 32-bit app and 64-bit device pointers (or vice versa)?